hw/mcu/stm/stm32_common/src/hal_os_tick.c

```diff
@@ -54,6 +55,315 @@ __WFI(void)
+            sub_seconds &= ~((1 << sub_seconds_tick_bits) - 1);
+            /* Add requested number of ticks. */
+            sub_seconds += (ticks & (OS_TICKS_PER_SEC - 1)) << sub_seconds_tick_bits;
+            /* Remove any overflow which is OK since we are sticking to 1 second limit. */
+            sub_seconds &= (1 << SUB_SECONDS_BITS) - 1;
+            /* Convert back to down-counting sub-seconds for alarm to use. */
+            alarm.AlarmTime.SubSeconds = alarm.AlarmTime.SecondFraction - sub_seconds;
+            alarm.AlarmMask = RTC_ALARMMASK_ALL;
+        } else {
+            sub_seconds = rtc_time_to_sub_seconds(&alarm.AlarmTime);
+            sub_seconds += ticks << sub_seconds_tick_bits;
+            sub_seconds_to_rtc(sub_seconds, &alarm.AlarmTime);
+            alarm.AlarmMask = RTC_ALARMMASK_DATEWEEKDAY;
+        }
+        HAL_RTC_SetAlarm_IT(&rtc, &alarm, RTC_FORMAT_BIN);
+    }
+
+    __DSB();
+    __WFI();
+
+    if (ticks > 0) {
+        rtc_update_time();
+    }
+}
+
+/* ST HAL interrupt handler calls this function. */
+void
+HAL_RTC_AlarmAEventCallback(RTC_HandleTypeDef *hrtc)
+{
+    (void)hrtc;
+    rtc_update_time();
+}
+
+void
+RTC_Alarm_IRQHandler(void)
+{
+    int sr;
+
+    os_trace_isr_enter();
+
+    OS_ENTER_CRITICAL(sr);
+    HAL_RTC_AlarmIRQHandler(&rtc);
+    OS_EXIT_CRITICAL(sr);
+
+    os_trace_isr_exit();
+}
+
+static void
+stm32_rtc_os_time_change(const struct os_time_change_info *info, void *arg)
+{
+    struct clocktime ct;
+    RTC_DateTypeDef date;
+    int32_t sub_seconds;
+    int sr;
+
+    timeval_to_clocktime(info->tci_cur_tv, NULL, &ct);
+    date.Year = ct.year - 2000;
+    date.Month = ct.mon;
+    date.Date = ct.day;
+    /* ST HAL Sunday is 7. */
+    date.WeekDay = ct.dow ? ct.dow : 7;
+    sub_seconds = ct.usec * 64 / 15625;
+
+    OS_ENTER_CRITICAL(sr);
+
+    alarm.AlarmTime.Hours = ct.hour;
+    alarm.AlarmTime.Minutes = ct.min;
+    alarm.AlarmTime.Seconds = ct.sec;
+    alarm.AlarmTime.SubSeconds = alarm.AlarmTime.SecondFraction;
+
+    HAL_RTC_SetTime(&rtc, &alarm.AlarmTime, RTC_FORMAT_BIN);
+    HAL_RTC_SetDate(&rtc, &date, RTC_FORMAT_BIN);
+    if (sub_seconds) {
+        RTC->SHIFTR = RTC_SHIFTR_ADD1S | sub_seconds;
+    }
+    last_rtc_time = rtc_time_to_sub_seconds(&alarm.AlarmTime);
+    alarm.AlarmTime.SubSeconds -= sub_seconds_per_tick;
+    alarm.AlarmMask = RTC_ALARMMASK_ALL;
+    HAL_RTC_SetAlarm_IT(&rtc, &alarm, RTC_FORMAT_BIN);
+
+    OS_EXIT_CRITICAL(sr);
+}
+
+static struct os_time_change_listener rtc_setter = {
+    .tcl_fn = stm32_rtc_os_time_change,
+};
+
+static void
+set_os_datetime_from_rtc(const RTC_TimeTypeDef *time, const RTC_DateTypeDef *date)
+{
+    struct os_timeval utc;
+    struct clocktime ct;
+    ct.year = 2000 + date->Year;
+    ct.mon = date->Month;
+    ct.day = date->Date;
+    ct.hour = time->Hours;
+    ct.min = time->Minutes;
+    ct.sec = time->Seconds;
+    ct.usec = ((time->SecondFraction - time->SubSeconds) * 15625) >> SUB_SECONDS_BITS;
```

Comment:
ct.dow is left uninitialized here

Reply:
ct.dow is not used for conversions from clocktime, it is filled when conversion is done other way round.
day of week can be incorrect in hardware, MCU will not enforce correct value so there is no guarantee that value extracted from registers is correct.
I saw no point in computing correct value since conversion function will not use it any way.